README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

<!-- badges: start -->

![GitHub top
language](https://img.shields.io/github/languages/top/InseeFrLab/fmtsas)
[![R check
status](https://github.com/InseeFrLab/fmtsas/workflows/R-CMD-check/badge.svg)](https://github.com/InseeFrLab/fmtsas/actions/workflows/check-release.yaml)
[![codecov](https://codecov.io/gh/InseeFrLab/fmtsas/branch/master/graph/badge.svg?token=25MHI8O62M)](https://codecov.io/gh/InseeFrLab/fmtsas)
[![CRAN
status](https://www.r-pkg.org/badges/version/fmtsas)](https://cran.r-project.org/package=fmtsas)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

<!-- ![](https://gitlab.insee.fr/uploads/-/system/project/avatar/1134/visuel_fmt_sas.png?width=64) -->

# fmtsas

Le package R **fmtsas** permet d’importer dans R des données servant à
construire des formats SAS.

Pour cela, il construit une liste de vecteurs, à partir de tables SAS ou
de programmes SAS. Cette liste pourra ensuite être utilisée pour
effectuer des conversions ou agrégations.

*Le package ne gère (pour l’instant) que des formats caractères
comportant un nombre défini de modalités.*

## Installation

``` r
# install.packages("remotes")
remotes::install_github("InseeFrLab/fmtsas")
```

## Chargement

``` r
library(fmtsas)
```

## Exemples d’import des données

### Depuis une table : `from_tab`

En entrée un data.frame qui proviendra typiquement d’une table SAS
servant dans une `proc format` avec l’option `CNTLIN=`.

``` r
format_data <- 
  data.frame(
    FMTNAME = c( "fmt1_", "fmt1_", "fmt1_", "fmt1_",  "sexe",  "sexe"),
    TYPE    = c(     "C",     "C",     "C",     "C",     "C",     "C"),
    START   = c(     "A",     "B",     "C",      NA,     "1",     "2"),
    LABEL   = c(     "A",    "BC",    "BC", "ERROR", "Homme", "Femme"),
    HLO     = c(      NA,      NA,      NA,     "O",      NA,      NA)
  )

conv_t <- from_tab(format_data)

conv_t
#> $fmt1_
#> # a character `fmtsas` object
#>    A    B    C 
#>  "A" "BC" "BC" 
#> [other] "ERROR"
#> $sexe
#> # a character `fmtsas` object
#>       1       2 
#> "Homme" "Femme"
```

### Depuis un programme : `from_pgm`

En entrée un programme SAS contenant une ou plusieurs `proc format` avec
`value(s)`.

``` r
test_pgm <-
  'PROC FORMAT;
     VALUE $ sexe "1"="Homme" "2"="Femme" ;
     value vnum
       0-99 = "petit" 100-high = "grand" ; 
     value $rega /* (geographie) */
       "01", "02", "03", "04", "06" = "Outre-mer"
       "11", "24", "27", "28", "32",
       "44", "52", "53", "75", "76",
       "84", "93", "94"             = "Métropole"
       other                        = "ERREUR" ;
   RUN;'

conv_p <- from_pgm(test_pgm)
#> Warning in from_pgm(test_pgm): 
#> Format(s) numerique(s) ignore(s) :
#>   vnum

conv_p
#> $sexe
#> # a character `fmtsas` object
#>       1       2 
#> "Homme" "Femme" 
#> 
#> $rega
#> # a character `fmtsas` object
#>          01          02          03          04          06          11 
#> "Outre-mer" "Outre-mer" "Outre-mer" "Outre-mer" "Outre-mer" "Métropole" 
#>          24          27          28          32          44          52 
#> "Métropole" "Métropole" "Métropole" "Métropole" "Métropole" "Métropole" 
#>          53          75          76          84          93          94 
#> "Métropole" "Métropole" "Métropole" "Métropole" "Métropole" "Métropole" 
#> [other] "ERREUR"
```

## Utilisation après import

### Format de l’objet

Les fonctions `from_tab` et `from_pgm` retournent toutes deux une liste
de vecteurs :

  - les noms de la liste correspondent aux noms des formats SAS ;
  - les éléments de la liste sont des vecteurs contenant les relations
    entre valeurs initiales et valeurs converties ;
  - chaque élément a un éventuel attribut `"other"` ;
  - chaque élément est un objet de type `fmtsas_c`, ce qui permet
    notamment d’utiliser `[]` avec prise en compte des valeurs par
    défaut (`other`). Cf. infra pour plus de détails sur cet opérateur.

### Exemple d’utilisation

Par exemple, si l’on a un jeu de données avec des codes pour lesquels on
souhaite avoir une correspondance en clair et/ou une agrégation.

``` r
donnees <-
  data.frame(
    ID   = c("001", "002", "003", "004", "005"),
    SEXE = c(  "1",   "2",   "1",   "2",   "#"),
    REG  = c( "04",  "11",  "##",    NA,  "24")
  )
```

On utilise les listes générées par les fonctions `from_*` et l’opérateur
`[]` pour objet “fmtsas” :

``` r
donnees$SEXE_LIB <- conv_t$sexe[donnees$SEXE]
donnees$REG2 <- conv_p$rega[donnees$REG]

donnees
#>    ID SEXE  REG SEXE_LIB      REG2
#> 1 001    1   04    Homme Outre-mer
#> 2 002    2   11    Femme Métropole
#> 3 003    1   ##    Homme    ERREUR
#> 4 004    2 <NA>    Femme    ERREUR
#> 5 005    #   24        # Métropole
```

> Utiliser `[]` sur un élément d’une liste créée par `from_tab` ou
> `from_pgm` fonctionne comme une sélection sur des vecteurs *normaux*,
> à la différence que d’éventuels `NA` seront remplacés par :
> 
>   - la valeur contenue dans l’attribut “other”, si cet attribut est
>     présent ;
>   - la valeur initiale, s’il n’y a pas d’attribut “other”.
> 
> Ces modifications ont pour but de reproduire le comportement d’un
> format SAS.
> 
> Pour que cet opérateur dédié soit utilisé, il faut que **le package
> soit chargé**. Dans le cas contraire, l’opérateur `[]` de base sera
> utilisé.

### Conversion de code SAS

Le package offre également la possibilité de rechercher dans un
programme SAS les instructions qui créent des variables à partir de
formats (de la forme `NEW = put(OLD, $var.);`), pour générer le code R
correspondant :

``` r
pgm_sas <-
  "data t2 ; set t1 ;
     SEXE_LIB = put(SEXE, $sexe.) ;
     REG2 = put(REG, $rega.) ;
   run ;"

convert_put(pgm_sas, fmt_list = "conv_p", style = "dplyr")
#> mutate(
#>   SEXE_LIB = conv_p$sexe[SEXE],
#>   REG2 = conv_p$rega[REG]
#> )
convert_put(pgm_sas, fmt_list = "conv_p", style = "base")
#> <donnees>$SEXE_LIB <- conv_p$sexe[<donnees>$SEXE]
#> <donnees>$REG2 <- conv_p$rega[<donnees>$REG]
```

Il ne reste plus qu’à copier ces instructions dans un programme R (après
quelques éventuels ajustements).


NEWS.md
---
title: Package ‘fmtsas’
subtitle: Historique des modifications
output: rmarkdown::html_vignette
---

## fmtsas 0.1.3

[06/2021]

* migration sur github InseeFrLab

## fmtsas 0.1.2

[12/2020]

* modification username gitlab

## fmtsas 0.1.1

[12/2020]

* méthode `print.fmtsas_c`
* améliore messages
* dépôt migré vers gitlab.insee.fr

## fmtsas 0.1.0

[05/2020]

* fonctions d'import : `from_pgm`, `from_tab`
* fonctions de conversion : `convert_put`, `[.fmtsas_c`
* constructeur S3 de format caractères : `fmtsas_c`


tests/testthat.R
library(testthat)
library(fmtsas)

test_check("fmtsas")


tests/testthat/test_util.R
# :::rm_sas_comments ---------------------------------------------------------

context(":::rm_sas_comments")

test_that("type `/* */`", {

  expect_equal(
    rm_sas_comments("options /*1234*/mprint ;"),
    "options mprint ;"
  )

  # multiligne
  expect_equal(
    rm_sas_comments("options /*12\n34*/mprint ;"),
    "options mprint ;"
  )

  # plusieurs (lazy)
  expect_equal(
    rm_sas_comments("options /*1234*/mprint /*567*/;"),
    "options mprint ;"
  )

  # commentaire final (sans fermeture)
  expect_equal(
    rm_sas_comments("options mprint ;/*1234"),
    "options mprint ;"
  )

})

test_that("type `* ;`", {

  expect_equal(
    rm_sas_comments("*1234 ;options mprint ;"),
    "options mprint ;"
  )

  # plusieurs
  expect_equal(
    rm_sas_comments("*1234;options mprint ;*567; run ;"),
    "options mprint ; run ;"
  )

  # multiligne
  expect_equal(
    rm_sas_comments("*12\n34 ;options mprint ;"),
    "options mprint ;"
  )

  # commentaire final (sans fermeture)
  expect_equal(
    rm_sas_comments("options mprint ;*1234"),
    "options mprint ;"
  )

  expect_equal(
    rm_sas_comments('*value "a*b"="x";'),
    ""
  )

  expect_equal(
    rm_sas_comments('*value "a*b"="x";options mprint ;'),
    "options mprint ;"
  )

})

test_that("mix types", {

  expect_equal(
    rm_sas_comments("*1234 ;options mprint ;/*567*/ run ;"),
    "options mprint ; run ;"
  )

  # imbrication (1)
  expect_equal(
    rm_sas_comments("/*1234 ;*567 ;89*/options mprint ;"),
    "options mprint ;"
  )

  # imbrication (2)
  expect_equal(
    rm_sas_comments("*1234 /*options mprint*/;options mprint ;"),
    "options mprint ;"
  )

})


# :::switch_quote ---------------------------------------------------------

context(":::switch_quote")

test_that("switch_quote", {

  expect_equal(
    switch_quote("'a'b\"c\""),
    "\"a\"b'c'"
  )

  expect_equal(
    switch_quote("'a\"b\"c'"),
    "\"a'b'c\""
  )

})


tests/testthat/test_from_pgm.R
# :::value_to_vect --------------------------------------------------------

context(":::value_to_vect")

test_that("pas de couples", {

  expect_null(value_to_vect(''))

  expect_null(value_to_vect('a'))

  expect_null(value_to_vect('0-4 = "N"')) # numérique

  expect_null(value_to_vect('A-D = "AD"')) # syntaxe pas gérée

})

test_that("un seul =", {

  expect_equal(
    value_to_vect('"a"="A"'),
    fmtsas_c(c("a" = "A"))
  )

  # avec blancs surnuméraires
  expect_equal(
    value_to_vect('  "a"\n =\n  "A" '),
    fmtsas_c(c("a" = "A"))
  )

  # virgule
  expect_equal(
    value_to_vect('"a1","a2"="A"'),
    fmtsas_c(c("a1" = "A", "a2" = "A"))
  )

  # tiret
  expect_equal(
    value_to_vect('"a1"-"a2"="A"'),
    fmtsas_c(c("a1" = "A", "a2" = "A"))
  )

})

test_that("plusieurs =", {

  expect_equal(
    value_to_vect('"a"="A" "b"="B"'),
    fmtsas_c(c("a" = "A", "b" = "B"))
  )

  # other
  expect_equal(
    value_to_vect('"a1","a2"="A" other = "X"'),
    fmtsas_c(c("a1" = "A", "a2" = "A"), other = "X")
  )

  # virgules, tirets
  expect_equal(
    value_to_vect('"a1","a2"="A" "b1"-"b3"="B" "c"="C"'),
    fmtsas_c(c("a1" = "A", "a2" = "A", "b1" = "B", "b3" = "B", "c" = "C"))
  )

  # virgules, tirets + blancs intercalés + other
  expect_equal(
    value_to_vect('
      "a1", "a2" = "A"
      "b1"-"b3"  = "B"
      "c"        = "C"
      other      = "ERROR"
    '),
    fmtsas_c(
      c(
        "a1" = "A", "a2" = "A",
        "b1" = "B", "b3" = "B",
        "c" = "C"
      ),
      other = "ERROR"
    )
  )

})


# from_pgm ----------------------------------------------------------------

context("from_pgm")

test_that("exemple complet", {

  test_pgm <-
    'options mprint ;
     PROC FORMAT;
       VALUE $ sexe "1"="Homme" "2"="Femme" ;
       vALue vnum
         0-99 = "petit" 100-high = "grand" ;
       value $rega /* (geographie) */
         "11", "24" = "Métropole"
         "01", "02",
         "03"       = "Outre-mer" ;
       value $az "A"-"Z" = "AZ" ;
       value $az_ A-Z = "AZ" ;
     RUN;'

  expect_warning(
    res <- from_pgm(test_pgm),
    "numerique.+vnum$"
  )

  expect_warning(
    res <- from_pgm(test_pgm),
    "intervalles caracteres.+az$"
  )

  expect_equal(
    res,
    list(
      sexe = fmtsas_c(c("1" = "Homme", "2" = "Femme")),
      rega = fmtsas_c(c(
        "11" = "Métropole", "24" = "Métropole",
        "01" = "Outre-mer", "02" = "Outre-mer", "03" = "Outre-mer"
      )),
      az = fmtsas_c(c("A" = "AZ", "Z" = "AZ"))
    )
  )

})

test_that("quote", {

  test_pgm <-
    "PROC FORMAT;
       VALUE $ sexe \"1\"=\"Homme\" \"2\"=\"Femme\" ;
       value $rega
         '11', '24' = 'Métropole'
         '01', '02', '03' = 'Outre-mer';
     RUN;"

  expect_equal(
    from_pgm(
      test_pgm,
      quote = "double" # [+]
    ),
    list(sexe = fmtsas_c(c("1" = "Homme", "2" = "Femme")))
  )

  expect_equal(
    from_pgm(
      test_pgm,
      quote = "simple" # [+]
    ),
    list(
      rega = fmtsas_c(c(
        "11" = "Métropole", "24" = "Métropole",
        "01" = "Outre-mer", "02" = "Outre-mer", "03" = "Outre-mer"
      ))
    )
  )

})

test_that("other", {

  test_pgm <- 'value $sexe "1"="Homme" "2"="Femme" other = "?" ;'

  expect_equal(
    from_pgm(test_pgm),
    list(sexe = fmtsas_c(c("1" = "Homme", "2" = "Femme"), other = "?"))
  )

  expect_equal(
    from_pgm(
      paste(
        test_pgm,
        'VALUE $sexe_ "1"="Homme" "2"="Femme" OTHER="Inconnu";' # [+]
      )
    ),
    list(
      sexe  = fmtsas_c(c("1" = "Homme", "2" = "Femme"), other = "?"),
      sexe_ = fmtsas_c(c("1" = "Homme", "2" = "Femme"), other = "Inconnu")
    )
  )

})

test_that("source", {

  test_pgm <-
    c(
      "VALUE $ sexe \"1\"=\"Homme\" \"2\"=\"Femme\" ;",
       "value $rega",
       "  '11', '24' = 'Métropole'",
       "  '01', '02', '03' = 'Outre-mer';"
    )

  expect_equal(
    from_pgm(
      test_pgm,
      source = TRUE # [+]
    ),
    structure(
      list(sexe = fmtsas_c(c("1" = "Homme", "2" = "Femme"))),
      source = paste(test_pgm, collapse = "\n")
    )
  )

})

test_that("vide", {

  expect_warning(
    from_pgm(character(0)),
    "aucun format .de type caractere."
  )

  expect_warning(
    from_pgm(""),
    "aucun format .de type caractere."
  )

  expect_warning(
    from_pgm("data t2 ; set t1 ; run"),
    "aucun format .de type caractere."
  )

  expect_warning(
    from_pgm('value $az_ A-Z = "AZ" ;'),
    "aucun format .de type caractere."
  )

})

test_that("doublons", {

  test_pgm <-
    c(
      "VALUE $ sexe \"1\"=\"Homme\" \"2\"=\"Femme\" ;",
      "value $rega",
       "  '11', '24' = 'Métropole'",
       "  '01', '02', '03' = 'Outre-mer';",
      "VALUE $ sexe \"1\"=\"Homme\";"
    )

  expect_warning(
    from_pgm(test_pgm),
    "doublon"
  )

})


R/from_tab.R
# ..............................................................................
# Specifications
# https://documentation.sas.com/?docsetId=proc&docsetTarget=n1e19y6lrektafn1kj6nbvhus59w.htm&docsetVersion=9.4
# ..............................................................................

#' Convertit une table pour format SAS
#'
#' Convertit les données utilisées pour une proc format (option CNTLIN=) en une
#' liste de vecteurs qui pourra servir à effectuer des conversions.
#'
#' La table en entrée est un data.frame, à importer dans R à partir d'une table
#' SAS (par exemple avec le package **haven**).
#'
#' Cette fonction ne gère pour l'instant que des formats caractère. De plus,
#' s'il y a une colonne `END` dans la table en entrée, elle doit être égale à la
#' colonne `START`.
#'
#' La modalité SAS `other` (valeur par défaut), présente dans la table si une
#' colonne `HLO` existe et vaut `"O"`, est sauvegardée dans l'attribut
#' "other" pour les éléments de la liste concernés.
#'
#' @param sas_data un data.frame importé depuis une table SAS pour proc format.
#'   Ce data.frame doit contenir a minima les colonnes `FMTNAME`, `START`,
#'   `LABEL` et `TYPE` (majuscule ou minuscule).
#'
#' @return Une liste contenant autant d'éléments que de formats si les données
#'   avaient été générées par une proc format.
#'   - les noms de la liste correspondent aux noms des formats (`FMTNAME`) ;
#'   - les éléments de la liste sont des vecteurs contenant les relations entre
#'   valeurs initiales et valeurs converties ;
#'   - chaque élément a un éventuel attribut `"other"` ;
#'   - chaque élément est un objet de type [`fmtsas_c`], ce qui permet
#'     d'utiliser l'[opérateur de sélection][extract.fmtsas_c] avec prise en
#'     compte d'une valeur par défaut (`other`).
#'
#'   Voir les exemples pour l'utilisation de cette liste.
#'
#' @importFrom stats setNames
#' @export
#'
#' @seealso [from_pgm] pour importer les formats contenus dans un programme SAS.
#'
#' @examples
#' ## Donnees en entree (provenant d'une table SAS) :
#'
#' format_data <-
#'   data.frame(
#'     FMTNAME = c( "fmt1_", "fmt1_", "fmt1_", "fmt1_",  "sexe",  "sexe"),
#'     TYPE    = c(     "C",     "C",     "C",     "C",     "C",     "C"),
#'     START   = c(     "A",     "B",     "C",      NA,     "1",     "2"),
#'     LABEL   = c(     "A",    "BC",    "BC", "ERROR", "Homme", "Femme"),
#'     HLO     = c(      NA,      NA,      NA,     "O",      NA,      NA)
#'   )
#'
#' ## Conversion des formats :
#'
#' conv <- from_tab(format_data)
#' conv
#'
#' ## Utilisation :
#'
#' # soit un jeu de donnees contenant des codes a convertir en libelles
#' donnees <-
#'   data.frame(
#'     VAR1_CODE = c("A", "B", "A", "C", "Z",  NA),
#'     SEXE_CODE = c("1", "2", "1", "2",  NA, "Z")
#'   )
#'
#' # pour remplacer les codes par les libelles (pour VAR1)
#' donnees$VAR1_LIB  <- conv$fmt1_[donnees$VAR1_CODE]
#' donnees$VAR1_LIB2 <- conv$fmt1_[donnees$VAR1_CODE, keep_na = TRUE]
#' donnees$SEXE_LIB  <- conv$sexe[donnees$SEXE_CODE]
#'
#' donnees

from_tab <- function(sas_data) {

  stopifnot(is.data.frame(sas_data))

  # transforme colonne facteurs en caractere
  fct <- vapply(sas_data, is.factor, TRUE)
  sas_data[fct] <- lapply(sas_data[fct], as.character)

  # noms en majuscule pour eviter tout probleme de casse
  names(sas_data) <- toupper(names(sas_data))

  # verifs
  needed <- c("FMTNAME", "START", "LABEL", "TYPE")
  if (any(miss <- !needed %in% names(sas_data))) {
    stop("\nColonne(s) manquante(s) : ", paste(needed[miss], collapse = ", "))
  }
  if ("END" %in% names(sas_data)) {
    start_diff_end <-
      with(
        sas_data,
        any(START[!is.na(START)] != END[!is.na(START)])
      )
    if (start_diff_end) stop("les colonnes START et END doivent etre egales")
  }

  # exclut formats numériques
  sas_data_chr <- sas_data[sas_data$TYPE == "C", ]
  if (nrow(sas_data_chr) != nrow(sas_data)) {
    warning("formats hors type caractere (C) exclus")
  }

  # other (HLO)
  fmtnames <- unique(sas_data_chr$FMTNAME)
  others <-
    setNames(
      rep(NA_character_, length(fmtnames)),
      fmtnames
    )
  others_hlo <- character(0)
  if (any(names(sas_data_chr) == "HLO")) {
    others_hlo <-
      with(
        sas_data_chr[with(sas_data_chr, !is.na(HLO) & HLO == "O"), ],
        setNames(LABEL, FMTNAME)
      )
  }
  others[names(others_hlo)] <- others_hlo

  # creation liste
  res <-
    with(
      sas_data_chr,
      tapply(
        setNames(as.vector(LABEL), START),
        FMTNAME,
        function(x) list(x[!is.na(names(x))]) #(!is.na retire possibles `other`)
      )
    )

  # ajouts attributs other
  lapply(
    setNames(names(res), names(res)),
    function(nm) {
      oth <- unname(others[nm])
      fmtsas_c(
        res[[nm]],
        other = if (is.na(oth)) NULL else oth
      )
    }
  )

}


R/fmtsas_class.R
# Constructeurs -----------------------------------------------------------

#' Objet de classe S3 "fmtsas_c"
#'
#' Construit un objet représentant en R un format caractère SAS, à partir de
#' ses composantes (vecteur d'associations et modalité par défaut).
#'
#' La classe S3 `"fmtsas_c"` représente un format caractère SAS. Cette classe
#' hérite de la classe `"fmtsas"`, qui représente tout type de format
#' (caractère, numérique...).
#'
#' Un objet `"fmtsas_c"` est un vecteur nommé contenant les relations entre
#' valeurs initiales et valeurs converties. Il possède en outre un éventuel
#' attribut "other", où est stockée une valeur par défaut. Cette valeur sera
#' utilisée lors d'une conversion avec
#' l'[opérateur de sélection][extract.fmtsas_c](`[]`).
#'
#' Le nom du format n'est pas stocké dans ce type d'objet. Il pourra être
#' spécifié comme un des noms d'une liste de `fmtsas`. Une telle liste sera en
#' pratique générée par [from_tab] ou [from_pgm].
#'
#' @param x vecteur caractère (dont tous les éléments sont nommés) représentant
#'   un format SAS (association _valeur(s) entrée = valeur sortie_).
#' @param other modalité par défaut (paramètre optionnel). Un vecteur caractère
#'   de longueur 1.
#'
#' @return Un objet de classe `"fmtsas_c"` (et donc `"fmtsas"`), c'est-à-dire un
#'   vecteur nommé possédant un éventuel attribut "other".
#'
#' @export
#'
#' @examples
#' fmtsas_c(c("A" = "LIBA", "B" = "LIB_B"))
#' fmtsas_c(c("A" = "LIBA", "B" = "LIB_B"), other = "??")

fmtsas_c <- function(x, other = NULL) {

  if (!is.character(x)) {
    stop("`x` doit etre un vecteur caractere")
  }
  if (is.null(names(x)) || any(names(x) == "") || anyNA(names(x))) {
    stop("tous les elements doivent avoir un nom")
  }
  if (!is.null(other) && (!is.character(other) || length(other) != 1L)) {
    stop("`other` doit etre un vecteur caractere de longueur 1")
  }

  structure(
    x,
    other = other,
    class = c("fmtsas_c", "fmtsas")
  )

}


# Getters -----------------------------------------------------------------

other <- function(x) UseMethod("other")

other.fmtsas <- function(x) attr(x, "other")


# Setters -----------------------------------------------------------------

`other<-` <- function(x, value) UseMethod("other<-")

`other<-.fmtsas` <- function(x, value) {

  err_msg <- "`other` doit etre un vecteur caractere de longueur 1"

  if (!is.null(value)) {
    if (length(value) != 1) stop(err_msg)
    if (is.na(value) && is.logical(value)) { # pour autoriser ecriture NA (lgl)
      value <- NA_character_
    }
    if (!is.character(value)) stop(err_msg)
  }

  attr(x, "other") <- value
  x

}


# print -------------------------------------------------------------------

#' @export

print.fmtsas_c <- function(x, ...) {

  cat("# a character `fmtsas` object\n")
  print.default(c(x)) # c() pour supprimer les attributs (hors noms)
  if (!is.null(other(x))) {
    cat("[other] \"", other(x), "\"", sep = "")
  }

}


R/from_pgm.R
# ..............................................................................
# Specifications
# https://documentation.sas.com/?docsetId=proc&docsetTarget=p1upn25lbfo6mkn1wncu4dyh9q91.htm&docsetVersion=9.4
# ..............................................................................

# couple_regex
#
# Expression régulière pour détecter une association de modalités dans une proc
# format.
#
# _Constante non exportée._
#
# Détectera :
#   - "A" = "LIBA"
#   - "A","B" = "LIBAB"
#   - "A"-"Z" = "LIBAZ"
#
# Insensible aux espace et sauts de lignes surnumeraires.

couple_regex <- '(?:other|(?:"[^"]*"\\s*[,-]\\s*)*"[^"]*")\\s*=\\s*"[^"]*"'

# Transforme un format (programme) en vecteur
#
# Transforme un format sous forme de programme SAS en vecteur nommé.
#
# _Fonction auxiliaire non exportée._
#
# Gère plusieurs valeurs à gauche séparées par virgules, mais pas d'intervalles
# caractères ("A"-"B") car contenant un nombre indéfini de modalités possibles.
# Seules les deux bornes sont conservées dans le résultat (comme si tiret
# remplacé par virgule).
#
# @param value_txt chaîne de texte de la forme `"1","2" = "x" "3" = "y"` (peut
#   contenir espaces surnuméraires, sauts de lignes, mais pas de commentaires)
#
# @return Un vecteur de type `c("1" = "x", "2" = "x", "3" = "y")`.

#' @importFrom stats setNames

value_to_vect <- function(value_txt) {

  # extract equivalences par regex (detecte plusieurs valeurs à gauche)
  equiv <- unlist(
    stringr::str_extract_all(
      value_txt,
      sprintf("(?i)(%s)", couple_regex)
    )
  )

  mbs <- strsplit(equiv, "\\s*=\\s*")
  if (!length(mbs)) return(NULL) # probablement format numerique

  outs <-
    stringr::str_match(
      sapply(mbs, function(x) x[2]),
      '"([^"]+)"'
    )[ , 2]

  ins <- vapply(mbs, function(x) x[1], character(1))

  is_other <- tolower(ins) == "other"
  if (any(is_other)) {
    other <- outs[is_other] # for attr(,"other")
    ins  <- ins[!is_other] # rm other from ins
    outs <- outs[!is_other] # rm other from outs
  } else {
    other <- NULL
  }

  ins <- stringr::str_match_all(ins, '"([^"]+)"')
  ins <- lapply(ins, function(x) x[ , 2])

  rep_outs <- rep(outs, lengths(ins))

  fmtsas_c(
    setNames(rep_outs, unlist(ins)),
    other = other
  )

}

#' Convertit un programme de formats SAS
#'
#' Convertit les données contenues dans les `proc format` d'un programme SAS en
#' une liste de vecteurs qui pourra servir à effectuer des conversions.
#'
#' La fonction ne recherche que les formats de type caractère (`value $nom`).
#' Les formats numériques sont ignorés (un message d'avertissement dresse une
#' liste de ces formats, s'ils sont présents).
#'
#' Le programme peut se présenter sous la forme d'une chaîne unique de
#' caractères mais aussi d'un vecteur de plusieurs chaînes (typiquement le
#' résultat de la lecture d'un fichier par [readLines]). Les commentaires,
#' espaces et sauts de lignes surnuméraires sont autorisés. La casse du code SAS
#' (majuscule ou minuscule) est sans importance. Le programme peut en outre
#' contenir autre chose que des `proc format`.
#'
#' La fonction détecte plusieurs valeurs séparées par des virgules à gauche du
#' signe `=`. En revanche, les intervalles de caractères (type `"A"-"Z"`) ne
#' sont pas gérés car ils contiennent un nombre indéfini de modalités possibles.
#' Les bornes de l'intervalle seront toutefois prises en compte (comme si
#' `"A","Z"` était écrit). La syntaxe sans les guillements (`A-Z`), également
#' permise par SAS, n'est **pas prise en compte** et ces intervalles seront
#' ignorés.
#'
#' Il est possible de choisir le type de guillemets (double ou simple) entourant
#' les valeurs. Avec un programme contenant un mélange de guillemets simples et
#' doubles, la fonction ne détectera qu'un type et pas l'autre.
#'
#' La modalité SAS `other` (valeur par défaut) est sauvegardée dans l'attribut
#' "other" pour chaque élément de la liste. Si le format SAS n'a pas de valeur
#' par défaut, l'attribut n'est pas présent.
#'
#' @param sas_pgm un programme SAS sous la forme d'un vecteur de chaînes de
#'   caractères.
#' @param quote type de guillemet. SAS autorise deux types de guillemets pour
#'   décrire une chaîne de caractères. La fonction suppose que des guillemets
#'   doubles sont utilisés ("). Dans le cas contraire ('), spécifier
#'   `quote = "simple"`.
#' @param source conserver le code SAS dans un attribut `"source"` de l'objet
#'   en sortie.
#'
#' @return Une liste contenant autant d'éléments que de formats si les données
#'   avaient été générées par SAS via une `proc format`.
#'   - les noms de la liste correspondent aux noms des formats (`value $...`) ;
#'   - les éléments de la liste sont des vecteurs contenant les relations entre
#'   valeurs initiales et valeurs converties ;
#'   - chaque élément a un éventuel attribut `"other"` ;
#'   - chaque élément est un objet de type [`fmtsas_c`], ce qui permet
#'     d'utiliser l'[opérateur de sélection][extract.fmtsas_c] avec prise en
#'     compte d'une valeur par défaut (`other`).
#'
#'   Voir les exemples pour l'utilisation de cette liste.
#'
#' @importFrom stats setNames
#' @export
#'
#' @seealso [from_tab] pour importer les formats contenus dans une table SAS.
#'
#' @examples
#' ## Import d'un programme SAS :
#'
#' test_pgm <- readLines(
#'   system.file("extdata/pgm_format_test.sas", package = "fmtsas"),
#'   encoding = "UTF-8"
#' )
#' cat(test_pgm, sep = "\n")
#'
#' ## Conversion des formats :
#'
#' conv <- from_pgm(test_pgm)
#' conv
#'
#' ## Utilisation :
#'
#' # soit un jeu de donnees contenant des codes a convertir en libelles
#' donnees <-
#'   data.frame(
#'     ACT_CODE = c( "B",   NA,  "C",  "P",  "W", "F"),
#'     REG_CODE = c("94", "04", "44", "09", "01",  NA)
#'   )
#'
#' # pour remplacer les codes par les libelles (pour ACT_CODE)
#' donnees$ACT_LIB  <- conv$a13_[donnees$ACT_CODE]
#' donnees$ACT_LIB2 <- conv$a13_[donnees$ACT_CODE, keep_na = TRUE]
#' donnees$REG_LIB  <- conv$reg[donnees$REG_CODE]
#'
#' donnees

from_pgm <- function(sas_pgm,
                     quote = c("double", "simple"),
                     source = FALSE) {

  quote <- match.arg(quote)

  # collapse si vecteur longueur > 1
  sas_pgm <- paste(sas_pgm, collapse = "\n")
  if (source) orig <- sas_pgm

  # intervertit simples et doubles guillemets si besoin
  if (quote == "simple") {
    sas_pgm <- switch_quote(sas_pgm)
  }

  # suppr comments
  sas_pgm_nocom <- rm_sas_comments(sas_pgm)

  # extrait noms des formats + couples separes par "="
  values <-
    stringr::str_match_all(
      sas_pgm_nocom,
      sprintf('(?i)value\\s+\\$\\s*(\\w+)\\s+((?:%s\\s*)+);', couple_regex)
    )[[1]]

  if (!length(values)) {
    quote_alt <- if (quote == "double") "simple" else "double"
    warning(
      "aucun format (de type caractere) dans sas_pgm : ",
      "essayer `quote = \"", quote_alt, "\"` ?")
  }

  fmtnames <- values[ , 2]
  couples  <- values[ , 3] # ensemble des couples (séparés plus loin)

  # applique value_to_vect a chaque "value"
  assoc <- lapply(couples, value_to_vect)

  # avertit numeriques (ignore)
  fmtnum <- stringr::str_match_all(sas_pgm_nocom, '(?i)value\\s+(\\w+)')[[1]]
  if (length(fmtnum)) {
    warning(
      "\nFormat(s) numerique(s) ignore(s) :\n  ",
      paste(fmtnum[ , 2], collapse = ", ")
    )
  }

  # avertit character range
  chr_range <- grepl('"[^"]*"\\s*-\\s*)*"[^"]*"\\s*=', couples)
  if (any(chr_range)) {
    warning(
      "\n",
      "Presence d'intervalles caracteres (\"A\"-\"Z\"). ",
      "Seules les bornes ont ete prises en compte :\n  ",
      paste(fmtnames[chr_range], collapse = ", ")
    )
  }

  # avertit doublons noms formats
  dupl_fmt <- duplicated(fmtnames)
  if (any(dupl_fmt)) {
    warning(
      "\nNoms de formats en doublon :\n  ",
      paste(fmtnames[dupl_fmt], collapse = ", ")
    )
  }

  res <- setNames(assoc, fmtnames)
  if (source) attr(res, "source") <- orig

  res

}
